Add flexibility to copy_from_home

esalberg opened the pull request:
- allow disabling the copy altogether
 - allow changing the file mode

esalberg commented:
Same as #217 - let me know if acceptance checks are needed, and I'm assuming the Beaker test will fail.  Also, these two PRs update the same files, so I can rebase one of them when/if the first is accepted.

esalberg commented:
Rebased for init.pp updates.

esalberg commented:
Similar to #217 
We were given an app-supplied Tomcat rpm, which our app team insisted on using because it was "supported" by the vendor. It actually ended up being mostly compatible with the module, except for having a strange directory listing (and some items I worked around in our profile). I yum installed the rpm to see what it did, then backfilled the info in the module so that the module could manage the service, connectors, etc. I didn't actually have docs from the vendor (sorry about that).

App: Symphony
Package: tomcat-8.0.41-25.x86_64.rpm (RHEL7)
catalina_home: /data/tomcat
catalina_base: /opt/tomcat

While the biggest issue is that they wanted the config files to have mode 0664 instead of 0660, I figured it was worthwhile adding the option not to copy the files as well.  I am copying the files, so I could revert that part of the change if you'd prefer.

hunner commented:
Related commit for reference: https://github.com/puppetlabs/puppetlabs-tomcat/pull/210/commits/e4ac84b187d8a1879e6d3c72cafd128749025e24

These parameters might be more conventional to be named `$manage_copy_from_home` (from this PR) and `$copy_from_home_list` (from the above-referenced commit)

lmayorga1980 commented on manifests/instance.pp:
```diff
@@ -35,6 +38,8 @@
   $use_jsvc               = undef,
   $use_init               = undef,
+  $copy_from_home         = undef,
+  $copy_from_home_mode    = undef,
```
Don't you need an extra variable that holds the files that you want to control/copy? There are scenarios where you don't want to copy all the default files.

hunner replied:
Yeah, it's missing here, and I wanted to add it.

esalberg commented:
Updated as follows:

Removed variables from init.pp.  Added defaults to instance.pp
Renamed variables to manage_copy_from_home, copy_from_home_list, and copy_from_home_mode.  Changed references to copy_to_base_list to copy_from_home_list for consistency (I didn't want to rename the instance/copy_from_home.pp file).

I looked at the referenced commit (to support log4j), but I don't see it in my latest master branch?  I incorporated / updated it anyway, but I wasn't sure if there could be regressing depending on where it was merged.

lmayorga1980 commented:
:+1:

hunner commented on spec/defines/instance_spec.rb:
```diff
@@ -189,4 +189,67 @@
+      {
+        catalina_home: '/opt/apache-tomcat',
+        catalina_base: '/opt/apache-tomcat/foo',
+        copy_from_home_list: '/opt/apache-tomcat/foo/conf/catalina.policy',
```
The code in #217 uses a loop to prefix the *relative* paths of `$dir_list` with the `catalina_base`. Should `copy_from_home_list` also take a list of relative paths to copy from `catalina_home`?

esalberg replied:
I don't object, but copy_from_home uses a private manifest to copy out the files that currently relies on the full path instead of relative.  
To rework this, I'd need to check if any other manifests besides instance.pp call it, then I'd either move the private manifest code into instance.pp, or take the dir_list code and move it into a matching private manifest (make it consistent one way or another).

hunner commented on manifests/init.pp:
```diff
@@ -27,18 +27,26 @@
 # @param manage_properties
 #   Boolean specifying whether or not to manage the catalina.properties file. Defaults to true.
+#
+# [*copy_from_home*]
```
The inline docs have been updated for eventually being used by puppet-strings documentation generation. You can see the format above.

esalberg replied:
Removed from init.pp altogether.  Do the inline docs for instance.pp need to be updated as well?

esalberg commented:
Code has been updated to move the parameters to tomcat::instance. I also rearranged the tomcat::instance section of the README.md file.

I did notice that the instance.pp file doesn't have parameters listed in Puppet Strings format, nor are the booleans typed. I didn't want to add that in as part of this PR specifically, but do you need a PR for that cleanup?

esalberg commented:
Rebased against master.
